Source/WebCore/Modules/speech/SpeechSynthesis.h

```diff
@@ -88,32 +89,39 @@ class SpeechSynthesis : public PlatformSpeechSynthesizerClient, public SpeechSyn
     void speakingErrorOccurred(PlatformSpeechSynthesisUtterance&) override;
     void boundaryEventOccurred(PlatformSpeechSynthesisUtterance&, SpeechBoundary, unsigned charIndex, unsigned charLength) override;
 
-    // SpeechSynthesisClient override methods
+    // SpeechSynthesisClientObserver
     void didStartSpeaking() override;
     void didFinishSpeaking() override;
     void didPauseSpeaking() override;
     void didResumeSpeaking() override;
     void speakingErrorOccurred() override;
     void boundaryEventOccurred(bool wordBoundary, unsigned charIndex, unsigned charLength) override;
     void voicesChanged() override;
-    
+
+    // ActiveDOMObject
+    const char* activeDOMObjectName() const final;
+    bool virtualHasPendingActivity() const final;
+
     void startSpeakingImmediately(SpeechSynthesisUtterance&);
     void handleSpeakingCompleted(SpeechSynthesisUtterance&, bool errorOccurred);
 
-    ScriptExecutionContext* scriptExecutionContext() const final { return ContextDestructionObserver::scriptExecutionContext(); }
+    // EventTarget
+    ScriptExecutionContext* scriptExecutionContext() const final { return ActiveDOMObject::scriptExecutionContext(); }
     EventTargetInterface eventTargetInterface() const final { return SpeechSynthesisEventTargetInterfaceType; }
     void refEventTarget() final { ref(); }
     void derefEventTarget() final { deref(); }
+    void eventListenersDidChange() final;
     
     PlatformSpeechSynthesizer& ensurePlatformSpeechSynthesizer();
     
     RefPtr<PlatformSpeechSynthesizer> m_platformSpeechSynthesizer;
-    Vector<Ref<SpeechSynthesisVoice>> m_voiceList;
+    std::optional<Vector<Ref<SpeechSynthesisVoice>>> m_voiceList;
     std::unique_ptr<SpeechSynthesisUtteranceActivity> m_currentSpeechUtterance;
     Deque<Ref<SpeechSynthesisUtterance>> m_utteranceQueue;
     bool m_isPaused;
     BehaviorRestrictions m_restrictions;
     WeakPtr<SpeechSynthesisClient> m_speechSynthesisClient;
+    std::atomic<bool> m_hasEventListener;
```

Comment:
Why does this need to be atomic?

Reply:
`virtualHasPendingActivity()` and `eventListenersDidChange()` are invoked on different threads (`virtualHasPendingActivity()` is on a non-main thread).

Reply:
Yeah but `virtualHasPendingActivity` gets called while the main thread is paused so I don't think there is a chance of data race here.

Reply:
I see; will update it